src/vsr/routing.zig

```diff
@@ -0,0 +1,877 @@
+        //
+        //   1 → 2
+        //  ↗
+        // ♔       3
+        //  ↘  ↗  ↙
+        //   5   4
+        //
+        // sum = (1 + 1) + (1 + 2 + 1)   (prepares, two replicas transposed)
+        //       + (1 + 2) + (1 + 2 + 3) (prepare ok)
+        //     = 15
+        //
+        // The difference between 14 and 15 is less than 10% but more than 5%.
+        //
+        // Why 10%: just feels like a reasonable number! median varies more than sum, so we need
+        // larger tolerance there.
+        inline for (.{
+            .{ "median", ratio(1, 10) },
+            .{ "maximum", ratio(1, 10) },
+            .{ "sum", ratio(1, 20) },
+        }) |field_threshold| {
+            const field, const threshold = field_threshold;
+            if (less_signifiantly(@field(lhs, field), @field(rhs, field), threshold)) return true;
+            if (@field(lhs, field).ns > @field(rhs, field).ns) return false;
+        }
+        return false;
+    }
+
+    // Returns true if lhs + lhs⋅threshold < rhs.
+    fn less_signifiantly(lhs: Duration, rhs: Duration, thershold: Ratio) bool {
+        assert(thershold.numerator < thershold.denominator);
+        return lhs.ns * (thershold.numerator + thershold.denominator) <
+            rhs.ns * thershold.denominator;
+    }
+
+    fn avg(lhs: Cost, rhs: Cost) Cost {
+        return .{
+            .median = .{ .ns = @divFloor(lhs.median.ns + rhs.median.ns, 2) },
+            .maximum = .{ .ns = @divFloor(lhs.maximum.ns + rhs.maximum.ns, 2) },
+            .sum = .{ .ns = @divFloor(lhs.sum.ns + rhs.sum.ns, 2) },
+        };
+    }
+
+    fn ewma_add(old: Cost, new: Cost) Cost {
+        return .{
+            .median = ewma_add_duration(old.median, new.median),
+            .maximum = ewma_add_duration(old.maximum, new.maximum),
+            .sum = ewma_add_duration(old.sum, new.sum),
+        };
+    }
+
+    fn ewma_add_duration(old: Duration, new: Duration) Duration {
+        return .{
+            .ns = @divFloor((old.ns * 4) + new.ns, 5),
+        };
+    }
+};
+
+const Routing = @This();
+
+pub fn init(options: struct {
+    replica: u8,
+    replica_count: u8,
+    standby_count: u8,
+}) Routing {
+    assert(options.replica < options.replica_count + options.standby_count);
+    assert(options.replica_count <= constants.replicas_max);
+    assert(options.standby_count <= constants.standbys_max);
+
+    const route = route_view_default(0, options.replica_count);
+
+    return .{
+        .replica = options.replica,
+        .replica_count = options.replica_count,
+        .standby_count = options.standby_count,
+
+        .view = 0,
+        .a = route,
+        .a_cost = null,
+
+        .b = route,
+        .b_cost = null,
+
+        .history = .{OpHistory.root()} ** history_max,
+    };
+}
+
+pub fn view_change(routing: *Routing, view: u32) void {
+    assert(view > routing.view or (view == 0 and routing.view == 0));
+    assert(routing.history_empty());
```

Comment:
The design is that the caller should explicitly call `.view_change` and `.history_reset`. `view_change` _can_ automagically reset the history, but I'd rather have this visible at the call-site.